bench/BenchIndices.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE BangPatterns        #-}
+-- |
+-- Copyright   : (c) 2020 Peter Duchovni
+-- License     : BSD3-style (see LICENSE)
+--
+-- Maintainer  : Peter Duchovni <caufeminecraft+github@gmail.com>
+--
+-- Benchmark elemIndex, findIndex, elemIndices, and findIndices
+module Main (main) where
+
+import           Data.Foldable                         (foldMap)
+import           Data.Monoid
+import           Data.String
+import           Gauge
+import           Prelude                               hiding (words)
+import           Data.Word                             (Word8)
+
+import qualified Data.ByteString                       as S
+import qualified Data.ByteString.Unsafe                as S
+
+
+------------------------------------------------------------------------------
+-- Benchmark
+------------------------------------------------------------------------------
+
+
+
+-- lines of 200 letters from a to e, followed by repeated letter f
+absurdlong :: S.ByteString
+absurdlong = S.replicate 200 0x61 <> S.singleton 0xa
+          <> S.replicate 200 0x62 <> S.singleton 0xa
+          <> S.replicate 200 0x63 <> S.singleton 0xa
+          <> S.replicate 200 0x64 <> S.singleton 0xa
+          <> S.replicate 200 0x65 <> S.singleton 0xa
+          <> S.replicate 999999 0x66
+
+main :: IO ()
+main = do
+  Gauge.defaultMain
+    [ bgroup "ByteString strict first index" $
+        [ bench "FindIndices" $ nf bench_find_indices_first absurdlong
+        , bench "ElemIndices" $ nf bench_elem_indices_first absurdlong
+        , bench "FindIndex"   $ nf bench_find_index_first   absurdlong
+        , bench "ElemIndex"   $ nf bench_elem_index_first   absurdlong
+        ]
+    , bgroup "ByteString strict second index" $
+        [ bench "FindIndices" $ nf bench_find_indices_second absurdlong
+        , bench "ElemIndices" $ nf bench_elem_indices_second absurdlong
+        , bench "FindIndex"   $ nf bench_find_index_second   absurdlong
+        , bench "ElemIndex"   $ nf bench_elem_index_second   absurdlong
+        ]
+    , bgroup "ByteString index equality inlining" $
+        [ bench "FindIndices/inlined"     $ nf bench_find_indices_inline   absurdlong
+        , bench "FindIndices/non-inlined" $ nf bench_find_indices_noinline absurdlong
+        , bench "FindIndex/inlined"       $ nf bench_find_index_inline     absurdlong
+        , bench "FindIndex/non-inlined"   $ nf bench_find_index_noinline   absurdlong
+        ]
+    ]
+
+safeHead :: [Int] -> Maybe Int
+safeHead (!x:_) = Just x
+safeHead _ = Nothing
+{-# INLINE safeHead #-}
```

Comment:
Use [`listToMaybe`](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Maybe.html#v:listToMaybe) please. Since you use `nf`, there is not much point in strict pattern here.